#summary Converted POD documentation.
#labels ManPage
_This page was generated from POD using [http://code.google.com/p/pod2gcw pod2gcw] and is not intended for editing._

==NAME==
bam2wig.pl

==SYNOPSIS==
bam2wig.pl `[`--options...`]` <**filename>

{{{
  Options:
  --in <filename>
  --out <filename> 
  --position [start|mid|span]
  --coverage
  --splice|split
  --pe
  --bin <integer>
  --shift
  --shiftval <integer>
  --sample <integer>
  --strand
  --qual <integer>
  --inter|fix
  --bed
  --rpm
  --log [2|10]
  --(no)track
  --bw
  --bwapp </path/to/wigToBigWig or /path/to/bedGraphToBigWig>
  --(no)gz
  --version
  --help
}}}
{{{
}}}
==OPTIONS==
The command line flags and descriptions:

__--in <**filename>

__
  Specify the input Bam alignment file. The file should be sorted and  indexed, although it may be indexed automatically
__
  __
__--out <**filename>

__
  Specify the output filename. By default it uses the base name of the  input file.
__
  __
__--position `[`start|mid|span`]`

__
  Specify the position of the alignment coordinate which should be  recorded. Several positions are accepted: the start (5') position of  the alignment, the midpoint of the alignment, or at all positions  along the length of the alignment (span). Note that the   span option gives coverage but not a true count of the number of  alignments, unlike start or mid. With paired-end alignments, the  positions are relative to the entire insert fragment defined by two  alignments. The default value is the start position for single-end and  midpoint position for paired-end alignments.
__
  __
__--coverage

__
  Quickly calculate the coverage of the alignments over the genome,  either at single bp resolution (default) or in bins. This method ignores  the position, quality, strand, shift, and log options. It uses a fast  low-level interface to the Bam file to eke out performance. It is  equivalent to specifying --position=span, --fix, --split,  --nope, --noshift, --nostrand, --qual=0, --norpm, and no log.
__
  __
__--splice

__
__--split

__
  The Bam file alignments may contain splices, where the  read is split between two separate alignments. This is most common  with splice junctions from RNA-Seq data. In this case, treat each  alignment as a separate tag. This only works with single-end alignments.  Splices are disabled for paired-end reads.
__
  __
__--pe

__
  The Bam file consists of paired-end alignments, and only properly  mapped pairs of alignments will be counted. The default is to  treat all alignments as single-end.
__
  __
__--bin <**integer>

__
  Specify the window or bin size in which alignments are counted. This  may make visualization in browsers easier when zoomed out to large  views. This option is compatible with all modes of counting, including  coverage. The default is to count at single basepair resolution.
__
  __
__--shift

__
  Specify that the positions of the alignment should be shifted towards  the 3' end. Useful for ChIP-Seq applications, where only the ends of  the fragments are counted and often seen as discrete peaks on separate  strands flanking the true target site. This option is disabled with  paired-end and spliced reads, and when --position=span is enabled.
__
  __
__--shiftval <**integer>

__
  Provide the value in bp that the record position should be shifted.  The value should be 1/2 the average length of the insert library  that was sequenced. The default is to automatically and empirically  determine the appropriate shift value. See below for the approach.
__
  __
__--sample <**integer>

__
  Indicate the number of top regions to sample when empirically  determining the shift value. The default is 100.
__
  __
__--strand

__
  Indicate that separate wig files should be written for each strand.  The output file basename is appended with either '`_`f' or '`_`r' for  both files. For paired-end RNA-Seq alignments that were generated with  !TopHat**, the XS attribute is honored for strand information.  The default is to take all alignments regardless of strand.
__
  __
__--qual <**integer>

__
  Set a minimum mapping quality score of alignments to count. The mapping quality score is a posterior probability that the alignment was mapped incorrectly, and reported as a -10Log10(P) value, rounded to the nearest integer (range 0..255). Higher numbers are more stringent. For performance reasons, when counting paired-end reads, only the left alignment is checked. The default value is 0 (accept everything).
__
  __
__--fix

__
__--inter

__
  Specify whether or not to record interpolating positions with a count  of 0. If true, a fixedStep wig file is written, otherwise a  variableStep wig file is written that only records the positions  where a tag is found. This will also work with bedGraph output.  This option is automatically enabled when the --bin option is enabled.  The default behavior is to not record empty positions.
__
  __
__--bed

__
  Specify whether or not to write a bedGraph (chromosome start stop value)  file or a traditional fixedStep or variableStep wiggle file. The  default is false.
__
  __
__--rpm

__
  Convert the data to Reads (or Fragments) Per Million mapped. This is useful  for comparing read coverage between different datasets. This conversion  is applied before converting to log, if requested. The default is no  conversion.
__
  __
__--log `[`2|10`]`

__
  Transform the count to a log scale. Specify the base number, 2 or  10. The counts are increased by 1 before taking a log transformation,  thus avoiding taking a log of 0. Only really useful with Bam alignment  files with high count numbers. Default is to not transform the count.
__
  __
__--(no)track

__
  Specify whether or not to include a track line in the wig file. This  option is automatically disabled when further converting to a bigWig  file.
__
  __
__--bw

__
  Specify whether or not the wig file should be further converted into  an indexed, compressed, binary !BigWig** file. The default is false.
__
  __
__--bwapp < /path/to/wigToBigWig or /path/to/bedGraphToBigWig >

__
  Specify the full path to Jim Kent's bigWig conversion utility. Two  different utilities may be used, bedGraphToBigWig or wigToBigWig,  depending on the format of the wig file generated. The wigToBigWig  is preferred only because smaller file sizes are produced, but with  slightly higher memory requirements. The application paths may be  set in the biotoolbox.cfg file.
__
  __
__--(no)gz

__
  Specify whether (or not) the output file should be compressed with  gzip. The default is compress the output unless a !BigWig** file is  requested.
__
  __
__--version

__
  Print the version number.
__
  __
__--help

__
  Display this POD documentation.
__
  __
==DESCRIPTION==
This program will enumerate aligned sequence tags and generate a wig,  or optionally !BigWig**, file. Each position in the wig file records the  number of alignments which map to that position. Alignments may be  counted at the start (5') or midpoint positions, or optionally  enumerated at every position across the alignment (resulting in a  coverage map rather than an alignment enumeration). Additionally,  alignments may be summed in bins (10, 100, 237, etc bp), which may  facilitate viewing in genome browsers zoomed out.

Both paired-end and single-end alignments may be counted. Alignments  with splices (e.g. RNA-Seq) may be counted singly or separately.  Alignment counts may be separated by strand, facilitating analysis of  RNA-Seq experiments.

For ChIP-Seq experiments, the alignment position may be shifted  in the 3' direction. This effectively merges the separate peaks  (representing the ends of the enriched fragments) on each strand  into a single peak centered over the target locus. The shift value  may be empirically determined from the sequencing data (see below).

The output wig file may be either a variableStep, fixedStep, or  bedGraph test format. The wig file may be further converted into a  compressed, indexed, binary bigWig format, dependent on the availability  of the appropriate conversion utilities.

More information about wiggle files can be found at  http://genome.ucsc.edu/goldenPath/help/wiggle.html, bedGraph at  http://genome.ucsc.edu/goldenPath/help/bedgraph.html, and bigWig at  http://genome.ucsc.edu/goldenPath/help/bigWig.html.

==RECOMMENDED SETTINGS==
The type of wig file to generate for your Bam sequencing file can vary  depending on your particular experimental application. Here are a few  common sequencing applications and my recommended settings for generating  the wig or bigWig file.

Straight coverage

__
  To generate a straight-forward coverage map, similar to what most genome  browsers display when using a Bam file as source, use the following  settings:
__
  __
{{{
 bam2wig.pl --coverage --in <bamfile>
}}}
Single-end ChIP-Seq

__
  When sequencing Chromatin Immuno-Precipitation products, one generally  is more interested in the number of tag counts, rather than coverage.  Hence, we can simply count the start position of the sequence tags.
__
  __
  To adjust the positions of tag count peaks to center over the presumed  site of interest, let the program empirically determine the shift  value from the sequence data (recommended). Otherwise, if you know  the mean size of your ChIP eluate fragments, you can use the --shiftval  option.
__
  __
  Finally, to compare ChIP-Seq alignments from multiple experiments,  convert your reads to Reads Per Million Mapped, which will help to  normalize read counts.
__
  __
{{{
 bam2wig.pl --pos start --shift --rpm --in <bamfile>
}}}
Paired-end ChIP-Seq

__
  If both ends of the ChIP eluate fragments are sequenced, then we do not  need to calculate a shift value. Instead, we will simply count at the  midpoint of each properly-mapped sequence pair.
__
  __
{{{
 bam2wig.pl --pos mid --pe --rpm --in <bamfile>
}}}
Unstranded RNA-Seq

__
  With RNA-Sequencing, we may be interested in either coverage (generating  a transcriptome map) or simple tag counts (differential gene expression),  so we can count in one of two ways.
__
  __
  To compare RNA-Seq data from different experiments, convert the read  counts to Reads Per Million Mapped, which will help to normalize read  counts.
__
  __
{{{
 bam2wig --pos span --rpm --in <bamfile>
 }}}
{{{
 bam2wig --pos mid --rpm --in <bamfile>
}}}
Stranded, single-end RNA-Seq

__
  If the library was generated in such a way as to preserve strand, then  we can separate the counts based on the strand of the alignment. Note  that the reported strand may be accurate or flipped, depending upon  whether first-strand or second-strand synthesized cDNA was sequenced,  and whether your aligner took this into account. Please check the  output wig files in a genome browser to verify which one is which, and  rename appropriately.
__
  __
{{{
 bam2wig --pos mid --strand --rpm --in <bamfile>
 }}}
{{{
 bam2wig --pos span --strand --rpm --in <bamfile>
}}}
Stranded, paired-end RNA-Seq

__
  Strand presents a complication when sequencing both ends of the cDNA  product from a library that preserves orientation. Currently,  the !TopHat** aligner can handle stranded, paired-end RNA-Seq alignments.  Because each pair will align to both strands, the aligner must record  separately which strand the original fragment should align. The !TopHat**  program records an 'XS' attribute for each alignment, and, if present,  bam2wig.pl will use this to set the strand.
__
  __
{{{
 bam2wig --pe --pos mid --strand --rpm --in <bamfile>
 }}}
{{{
 bam2wig --pe --pos span --strand --rpm --in <bamfile>
}}}
==SHIFT VALUE DETERMINATION==
To determine the shift value, the top enriched 500 bp regions (the  default number is 100) from the largest chromosome are identified by  their read coverage. Stranded read counts are then collected in 10 bp  bins over a 1.5 kb region encompassing the identified region. A  Pearson product-moment correlation coefficient is then reiteratively  determined between the stranded data as the bins are shifted from  30 to 500 bp. The shift corresponding to the highest R squared value  is retained for each sampled region, and the mean best shift for all  sampled regions is used as the final shift value. This approach works  best with clean, distinct peaks. Not all sampled regions may return  a significant R squared value. The peak shift may be evaluated by  viewing separate, stranded wig files together with the shifted wig  file in a genome browser.

==AUTHOR==
{{{
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112
}}}
This package is free software; you can redistribute it and/or modify it under the terms of the GPL (either version 1, or at your option, any later version) or the Artistic License 2.0.

