#summary Converted POD documentation.
#labels ManPage
_This page was generated from POD using [http://code.google.com/p/pod2gcw pod2gcw] and is not intended for editing._

==NAME==
get_features.pl

==SYNOPSIS==
get_features.pl --db <**text> `[`--options...`]`

{{{
  Options:
  --db <text>
  --feature <type | type:source>
  --all
  --sub
  --coord
  --start=<integer>
  --stop=<integer>
  --out <filename>
  --bed
  --gff 
  --(no)gz
  --version
  --help
}}}
==OPTIONS==
The command line flags and descriptions:

__--db <**text>

__
  Specify the name or SQLite file of the Bio::DB:!SeqFeature**::Store  database from which to collect the features. Other database schemas  are not currently supported. Required.
__
  __
__--feature <**type | type:source>

__
  Optionally specify the type of the feature(s) to collect. The GFF  primary_tag or primary_tag:source_tag should be specified. More than  one feature may be specified at a time, either as a comma de-limited  list or separate options. If not specified, then an interactive list  will be presented to the user for selection.
__
  __
__--all

__
  Optionally indicate that all features present in the database must  be included. By default, certain features may be excluded based on  parameters defined in the !BioToolBox** configuration file. See below  for details.
__
  __
__--sub

__
  Optionally include all subfeatures in the output. For example,  transcript, CDS, and/or exon subfeatures of a gene.
__
  __
__--coord

__
  When writing a standard format file, optionally include the chromosome,  start, stop, and strand coordinates. These are automatically included  when writing a BED or GFF format.
__
  __
__--start=<**integer>

__
__--stop=<**integer>

__
  Optionally specify adjustment values to adjust the reported start and  end coordinates of the collected regions. A negative value is shifted  upstream (5' direction), and a positive value is shifted downstream. Adjustments are made relative to the feature's strand. Adjustments  are ignored if a GFF file is written.
__
  __
__--out <**filename>

__
  Specify the output file name. Default is the joined list of features.
__
  __
__--bed

__
  Optionally indicate that a 6-column BED format file should be  written. Currently, 12-column BED formats with exon information is  not supported (yet).
__
  __
__--gff

__
  Optionally indicate that a GFF3 format file should be written. This  option enables the database features to be written completely with  all attributes. Coordinate adjustments are ignored.
__
  __
__--(no)gz

__
  Specify whether (or not) the output file should be compressed with gzip.
__
  __
__--version

__
  Print the version number.
__
  __
__--help

__
  Display this POD documentation.
__
  __
==DESCRIPTION==
This program will extract a list of features from a database and  write them out to a file. Specifically, the requested features in  a Bio::DB::!SeqFeature**::Store schema database are pulled and written  as either a list of named features with or without coordinate information,  a BED-style formatted file, or a GFF3-formatted file. The GFF option  is essentially a database dump, as it enables a low-level option to  write the features as original GFF3 lines complete with all attributes.

Features may be specified through their GFF type or primary_tag. They may be specified as a command-line option or selected interactively from a presented list. They may be restricted through two options defined in the !BioToolBox** configuration file, biotoolbox.cfg. These include a database-specific or default database option, "chromosome_exclude", which excludes features located on the listed chromosomes (such as the mitochondrial chromosome), and the "exclude_tags", which are attribute keys and values to be avoided. More information may be found in the configuration file itself.

==AUTHOR==
{{{
 Timothy J. Parnell, PhD
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112
}}}
This package is free software; you can redistribute it and/or modify it under the terms of the GPL (either version 1, or at your option, any later version) or the Artistic License 2.0.

