#summary Converted POD documentation.
#labels ManPage
_This page was generated from POD using [http://code.google.com/p/pod2gcw pod2gcw] and is not intended for editing._

==NAME==
average_gene.pl

==SYNOPSIS==
{{{
 average_gene.pl --db <database> --feature <text> --out <file> [--options]
 average_gene.pl --in <file> --out <file> [--options]
  }}}
{{{
  Options:
  --db <name|file.gff3>
  --feature [type, type:source, alias]
  --out <filename>
  --in <filename> 
  --data <dataset_name | filename>
  --method [mean|median|min|max|sum]
  --value [score|count|length]
  --bins <integer>
  --ext <integer>
  --extsize <integer>
  --min <integer>
  --strand [all|sense|antisense]
  --sum
  --smooth
  --set_strand
  --(no)log
  --help
}}}
==OPTIONS==
The command line flags and descriptions:

{{{
}}}
__--db <**name|file.gff3>

__
  Specify the name of the !BioPerl** !SeqFeature**::Store database to use as source. Alternatively, a single GFF3 file may be loaded into a in-memory database. Specifying the database is required for new feature data files. For pre-existing input data files, this value may be obtained from the input file metadata. However, if provided, it overrides the database listed in the file; this is useful for collecting data from multiple databases.
__
  __
__--out <**filename>

__
  Specify the output file name.
__
  __
__--in <**filename>

__
  Specify the filename of a data table containing the list of  features to average. The file must be in the 'tim_data' format and specify  a feature to use. If not specified, then a new feature list will be  generated from the database.
__
  __
__--feature `[`type, type:source, alias`]`

__
  Specify the type of feature from which to collect values. This is required  for new feature tables. Two types of values may be passed: either a specific  feature type present in the database, or an alias to one or more features.  The feature may be specified as either type or type:source. Aliases are  specified in the `tim_db_helper.cfg` file, and provide a shortcut to a  list of one or more features. More than feature may be included as a  comma-delimited list (no commas).
__
  __
__--data <**dataset_name | filename>

__
  Specify the name of the microarray data set for which you wish to  collect data. If not specified, the data set may be  chosen interactively from a presented list. When enumerating features,  the features' type or type:source values should be indicated. Alternatively, the name of a data file may be provided. Supported  file types include !BigWig** (.bw), !BigBed** (.bb), or single-end Bam  (.bam). The file may be local or remote.
__
  __
__--method `[`mean|median|min|max|sum`]`

__
  Specify the method of collecting and combining the data into each bin.  Three statistical methods for combining score values are allowed: mean,  median, minimum, maximum, and sum. The defined method does not affect  the interpolation or summary functions of the program, only initial data  collection. The default method is 'mean'.
__
  __
__--value `[`score|count|length`]`

__
  Optionally specify the type of data value to collect from the dataset or  data file. Three values are accepted: score, count, or length. The default  value type is score. Note that some data sources only support certain  types of data values. Wig and !BigWig** files only support score and count;  !BigBed** and database features support count and length and optionally  score; Bam files support basepair coverage (score), count (number of  alignments), and length.
__
  __
__--bins <**integer>

__
  Specify the number of bins that will be generated over the length  of the feature. The size of the feature is a percentage of the  feature length. The default number is 10, which results in bins of  size equal to 10% of the feature length.
__
  __
__--ext <**integer>

__
  Specify the number of extended bins on either side of the feature.  The bins are of the same size as determined by the feature  length and the --bins value. The default is 0.
__
  __
__--extsize <**integer>

__
  Specify the exact bin size in bp of the extended bins rather than using a percentage of feature of length.
__
  __
__--min <**integer>

__
  Specify the minimum feature size to be averaged. Features with a length below this value will not be skipped (all bins will have null values). This is to avoid having bin sizes below the average  microarray tiling distance.
__
  __
__--strand `[`all|sense|antisense`]`

__
  Specify whether stranded data should be collected. Three values are  allowed: all datasets should be collected (default), only sense  datasets, or only antisense datasets should be collected.
__
  __
__--sum

__
  Indicate that the data should be averaged across all features at each position, suitable for graphing. A separate text file will be written with the suffix '`_`summed' with the averaged data
__
  __
__--smooth

__
  Indicate that windows without values should (not) be interpolated from neighboring values. The default is false.
__
  __
__--set_strand

__
  For features that are not inherently stranded (strand value of 0),  impose an artificial strand for each feature (1 or -1). This will  have the effect of enforcing a relative orientation for each feature,  or to collected stranded data. This requires the presence a  column in the input data file with a name of "strand". Hence, it  will not work with newly generated datasets, but only with input  data files. Default is false.
__
  __
__--(no)log

__
  Dataset values are (not) in log2 space and should be treated  accordingly. Output values will be in the same space.
__
  __
__--help

__
  This help text.
__
  __
==DESCRIPTION==
This program will collect data across a gene body into bins to generate an 'average' gene summary. It is used to determine if there is a spatial  distribution preference for the dataset over gene bodies. The number of bins may be specified as a command argument (default 10). Additionally, extra bins  may be extended on either side of the gene (default 0 on either side). The bin  size is determined as a percentage of gene length.

The program writes out a tim data formatted text file.

{{{
}}}
==AUTHOR==
{{{
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112
}}}
This package is free software; you can redistribute it and/or modify it under the terms of the GPL (either version 1, or at your option, any later version) or the Artistic License 2.0.

{{{
}}}
{{{
}}}
