#summary A summary of the scripts available in BioToolBox
#labels Featured

=Main Programs=

List of programs included in the main CPAN distribution. These are organized by function.

===Data (score) collection===
	* [Pod_get_datasets get_datasets.pl] - Collect single data value for features
	* [Pod_get_relative_data get_relative_data.pl] - Collect data in bins flanking a specific point relative to a feature
	* [Pod_get_binned_data get_binned_data.pl] - Collect data in defined bins across a feature

===Data file manipulation===
	* [Pod_manipulate_datasets manipulate_datasets.pl] - Interactive column, row, and statistical manipulation of data files
	* [Pod_merge_datasets merge_datasets.pl] - Merge columns from two or more data files into one
	* [Pod_split_data_file split_data_file.pl] - Split a large data file into two or more smaller files based on criteria
	* [Pod_join_data_file join_data_file.pl] - Concatenate rows in two or more data files into one
	* [Pod_pull_features pull_features.pl] - Pull out rows from a data file based on a list of names

===Data analysis===
	* [Pod_graph_data graph_data.pl] - Generate xy plots between two columns of data
	* [Pod_graph_profile graph_profile.pl] - Generate profile graph across static positions 
	* [Pod_graph_histogram graph_histogram.pl] - Generate distribution histogram plots from columns of data
	* [Pod_data2frequency data2frequency.pl] - Convert column values to a frequency distribution
	* [Pod_run_cluster run_cluster.pl] - Generate k-means clusters for datasets
	* [Pod_change_cluster_order change_cluster_order.pl] - Change the order of k-means clusters from random to meaningful
	* [Pod_correlate_position_data correlate_position_data.pl] - Calculate positional correlation between two datasets 

===File format conversion===
	* [Pod_bam2wig bam2wig.pl] - Generate many styles of coverage maps based on sequence alignments
	* [Pod_data2wig data2wig.pl] - Convert any data table into a UCSC-style wig file
	* [Pod_data2bed data2bed.pl] - Convert any data table file into a UCSC-style bed file
	* [Pod_data2gff data2gff.pl] - Convert any data table file into a GFF file
	* [Pod_data2fasta data2fasta.pl] - Convert any data table file into a multi-fasta file
	* [Pod_wig2data wig2data.pl] - Convert any wig file into a standard data table file

===Obtaining genome annotation===
	* [Pod_db_setup db_setup.pl] - Quick and easy annotation database setup using UCSC annotation
	* [Pod_db_types db_types.pl] - Print a list of feature types in an annotation database
	* [Pod_ucsc_table2gff3 ucsc_table2gff3.pl] - Convert UCSC annotation tables into feature-rich GFF3 files
	* [Pod_get_ensembl_annotation get_ensembl_annotation.pl] - Export Ensembl annotation as feature-rich GFF3 files
	* [Pod_gff3_to_ucsc_table gff3_to_ucsc_table.pl] - Convert a GFF3 file to an UCSC annotation table

===Finding features and annotation===
	* [Pod_get_features get_features.pl] - Get a list of features from an annotation database
	* [Pod_get_feature_info get_feature_info.pl] - Get additional information for annotated features 
	* [Pod_get_gene_regions get_gene_regions.pl] - Get a list of implicit or calculated regions based on gene annotation
	* [Pod_get_intersecting_features get_intersecting_features.pl] - Find annotation in a database that intersects regions
	* [Pod_find_enriched_regions find_enriched_regions.pl] - Find regions across a genome whose score values exceed a threshold

===Sequence manipulation===
	* [Pod_filter_bam filter_bam.pl] - Filter a bam file of sequence alignments based on various criteria
	* [Pod_CpG_calculator CpG_calculator.pl] - Calculate observed and predicted CpG frequency over regions


=Extra programs=

Additional programs available through SVN only. These are specialized, out-of-date, or no longer useful scripts. Some people may still find these useful.

	* [Pod_bin_genomic_data bin_genomic_data.pl] - Collate genomic datasets into binned data
	* [Pod_compare_subfeature_scores compare_subfeature_scores.pl] - Compare the scores of subfeatures 
	* [Pod_bam2gff_bed bam2gff_bed.pl] - Covert Bam sequence alignments to a bed or gff file
	* [Pod_bar2wig bar2wig.pl] - Convert USeq or T2 Bar files to wig files
	* [Pod_map_transcripts map_transcripts.pl] - Identify transcript start and stop sites from transcriptome data
	* [Pod_ucsc_cytoband2gff3 ucsc_cytoband2gff3.pl] - Convert UCSC cytoband tables into GFF3 for GBrowse
	* [Pod_convert_yeast_genome_version convert_yeast_genome_version.pl] - Convert positional data between different yeast versions
	* [Pod_get_actual_nuc_sizes get_actual_nuc_sizes.pl] - Calculate actual nucleosome sizes from paired-end sequencing data
	* [Pod_get_bam_seq_stats get_bam_seq_stats.pl] - Calculate sequence stats from a Bam file
	* [Pod_novo_wrapper novo_wrapper.pl] - Pseudo parallel novoalign wrapper
	* [Pod_split_bam_by_isize split_bam_by_isize.pl] - Split a Bam file based on paired-end insert size
	* [Pod_intersect_nucs intersect_nucs.pl] - Intersect two sets of nucleosome files
	* [Pod_map_nucleosomes map_nucleosomes.pl] - Identify nucleosome positions from MNase sequencing
	* [Pod_verify_nucleosome_mapping verify_nucleosome_mapping.pl] - Verify mapped nucleosome positions with original data
	* [Pod_map_oligo_data2gff map_oligo_data2gff.pl] - Associate microarray oligo values with genomic positions
	* [Pod_process_microarray process_microarray.pl] - Quantile normalization of raw microarray data for Agilent, GenePix, and Nimblegen formats
	* [Pod_intersect_SNPs intersect_SNPs.pl] - Intersect SNPs in two or more VCF files to find common and unique
	* [Pod_locate_SNPs locate_SNPs.pl] - Locate SNPs relative to gene annotation and identify mutation impact
	* [Pod_big_file2gff3 big_file2gff3.pl] - Generate a GFF3 file and GBrowse conf stanzas for big files
	* [Pod_change_chr_prefix change_chr_prefix.pl] - Change a chromosome prefix for a variety of data file formats


