<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>bar2wig.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>bar2wig.pl</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>bar2wig.pl [--options...] &lt;filename&gt;</p>
<pre>

  Options:
  --in &lt;filename&gt; or &lt;directory&gt;
  --out &lt;filename&gt; 
  --barapp &lt;/path/to/Bar2Gr&gt;
  --method [mean | median | sum | max]
  --(no)track
  --(no)inter
  --bw
  --db &lt;database&gt;
  --chromof &lt;filename&gt;
  --bwapp &lt;/path/to/wigToBigWig&gt;
  --(no)gz
  --help</pre>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="in_filename" class="item">--in &lt;filename&gt;</a></strong></dt>

<dd>
<p>Specify the input bar file or (more likely) a directory of bar files. 
The bar files may be zipped (*.bar.zip).</p>
</dd>
<dt><strong><a name="out_filename" class="item">--out &lt;filename&gt;</a></strong></dt>

<dd>
<p>Specify the output filename. By default it uses the base name of the 
input file or the input directory name. The output file will either 
have a .wig or .bw file extension.</p>
</dd>
<dt><strong><a name="barapp_path_to_bar2gr" class="item">--barapp &lt;/path/to/Bar2Gr&gt;</a></strong></dt>

<dd>
<p>Specify the full path to David Nix's USeq or T2 application Bar2Gr (it 
is included in both software packages). By default it uses the path 
defined in the biotoolbox configuration file, tim_db_helper.cfg.</p>
</dd>
<dt><strong><a name="method_mean_median_sum_max" class="item">--method [mean | median | sum | max]</a></strong></dt>

<dd>
<p>Define the method used to combine multiple data values at a single 
position. Bar files can frequently have one or more values at a single 
position, usually due to either multiple identical oligo probes (microarray 
data) or multiple sequence tags aligning to the same position (next 
generation sequencing data). Typically, with microarray data, the values 
should be averaged (mean), while with sequencing data, values should be summed.
The default method is 'sum'.</p>
</dd>
<dt><strong><a name="no_track" class="item">--(no)track</a></strong></dt>

<dd>
<p>Wig files typically include a track line at the beginning of the file which 
defines the appearance. However, conversion of a wig file to bigWig 
requires that the track line is absent. Do not include the track line when 
generating bigWig files manually. The default is to include for wig files 
(true), and exclude for bigWig files (false).</p>
</dd>
<dt><strong><a name="no_inter" class="item">--(no)inter</a></strong></dt>

<dd>
<p>Bar files are typically in interbase (0-base) format, whereas wig files are 
typically 1-base format. Shift (or not) the position by 1 bp. Default is 
true.</p>
</dd>
<dt><strong><a name="bw" class="item">--bw</a></strong></dt>

<dd>
<p>Flag to indicate that a binary bigWig file should be generated rather than 
a text wig file.</p>
</dd>
<dt><strong><a name="db_database" class="item">--db &lt;database&gt;</a></strong></dt>

<dd>
<p>Specify the name of a Bio::DB database from which to extract chromosome 
names and sizes. This information is required when generating a bigWig 
file.</p>
</dd>
<dt><strong><a name="chromof_filename" class="item">--chromof &lt;filename&gt;</a></strong></dt>

<dd>
<p>Alternative to the --db argument, a pre-generated chromosome sizes text 
file may be specified. This text file should consist of two columns, 
delimited by whitespace, consisting of the chromosome name and size.</p>
</dd>
<dt><strong><a name="bwapp_path_to_wigtobigwig" class="item">--bwapp &lt;/path/to/wigToBigWig&gt;</a></strong></dt>

<dd>
<p>Specify the full path to Jim Kent's wigToBigWig conversion utility. By 
default it uses the path defined in the biotoolbox configuration file, 
tim_db_helper.cfg. If it is not defined here or in the config file, then 
the system path is searched for the executable. Finally, failing that, 
it will attempt to use Lincoln Stein's Bio::DB::BigFile module for 
conversion, if available.</p>
</dd>
<dt><strong><a name="no_gz" class="item">--(no)gz</a></strong></dt>

<dd>
<p>Specify whether (or not) the output wig file should be compressed with gzip.
This option does not affect bigWig files, but will affect intermediate 
wig files.</p>
</dd>
<dt><strong><a name="help" class="item">--help</a></strong></dt>

<dd>
<p>Display this POD documentation.</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program will convert binary .bar files to a conventional 
text-based wig file, with the option to further convert to a binary 
bigWig format.</p>
<p>The bar file is a Java serialized, floating-point, binary file used
natively by David Nix's TiMAT2 (T2) and USeq software packages. It allows
for effective visualization by the Integrated Genome Browser (IGB), but
very few (any?) other genome browsers.</p>
<p>The wig file is a commonly used text-based file format for exchanging and 
representing dense genomic data (microarray and sequencing) that is read 
by nearly all genome browsers. In most genome browsers, the data values are 
downsampled to 8-bit precision (0-255), sufficient for visualization but 
limited for data analysis.</p>
<p>The bigWig format is an indexed, compressed, binary extension of the wig
format. It maintains numeric precision, allows for rapid statistical
summaries, and can be rapidly and randomly accessed either locally or
remotely. It is supported by the UCSC and GBrowse genome browsers as well 
as biotoolbox scripts.</p>
<p>This program first uses David Nix's Bar2Gr application to convert the bar 
file to a very simplified text format, a .gr file. This is then further 
processed into one or more wig (or bigWig) files. Stranded data (denoted 
by _+_ and _-_ in the bar file names) is written to two stranded output 
files, appended with either '_f' or '_r' to the basename. The wig files 
are in variableStep format.</p>
<p>Conversion from wig to bigWig requires Jim Kent's wigToBigWig utility or 
Lincoln Stein's Bio::DB::BigFile support.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
