<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>generate_genomic_bins.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>generate_genomic_bins.pl</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>generate_genomic_bins.pl [--options...] --db &lt;database&gt; --out &lt;file&gt;</p>
<pre>

  Options:
  --db &lt;database&gt;
  --win &lt;integer&gt;
  --step &lt;integer&gt;
  --mito
  --split
  --max &lt;integer&gt;
  --out &lt;filename&gt; 
  --(no)gz
  --help</pre>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="db_database" class="item">--db &lt;database&gt;</a></strong></dt>

<dd>
<p>Specify the name of the BioPerl gff database to use as source. This is required.</p>
</dd>
<dt><strong><a name="win_integer" class="item">--win &lt;integer&gt;</a></strong></dt>

<dd>
<p>Optionally specify the window size. The default size is defined in the
configuration file, tim_db_helper.cfg.</p>
</dd>
<dt><strong><a name="step_integer" class="item">--step &lt;integer&gt;</a></strong></dt>

<dd>
<p>Optionally indicate the step size. The default is equal to the window size.</p>
</dd>
<dt><strong><a name="mito" class="item">--mito</a></strong></dt>

<dd>
<p>Specify whether the mitochondrial chromosome should be included. The 
default is false.</p>
</dd>
<dt><strong><a name="split" class="item">--split</a></strong></dt>

<dd>
<p>Specify whether the output file should be split into individual chromosome 
files.</p>
</dd>
<dt><strong><a name="max_integer" class="item">--max &lt;integer&gt;</a></strong></dt>

<dd>
<p>When splitting the output file, optionally specify the maximum number of 
lines in each file part.</p>
</dd>
<dt><strong><a name="out_filename" class="item">--out &lt;filename&gt;</a></strong></dt>

<dd>
<p>Specify the output filename. By default it uses</p>
</dd>
<dt><strong><a name="no_gz" class="item">--(no)gz</a></strong></dt>

<dd>
<p>Specify whether (or not) the output file should be compressed with gzip.</p>
</dd>
<dt><strong><a name="help" class="item">--help</a></strong></dt>

<dd>
<p>Display this POD documentation.</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program will generate a tim data format file of genomic bins or 
intervals. Generating bins of small size (say, 1 or 10 bp in size) 
for an entire genome, particularly metazoan genomes, produces 
extremely large data files, which demands large memory resources to 
load. This simple script avoids the memory demands by writing 
directly to file as the bins are generated. It will optionally also 
split the file by chromosome (using the biotoolbox script 
split_data_file.pl).</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
