<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>get_ensembl_annotation.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>get_ensembl_annotation.pl</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>get_ensembl_annotation.pl [--options...] --species &lt;text&gt;</p>
<pre>

  Options:
  --species &lt;text&gt;
  --out &lt;filename&gt;
  --(no)chromo
  --(no)protein
  --(no)rna
  --host &lt;host.address&gt;
  --user &lt;text&gt;
  --help</pre>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="species_text" class="item">--species &lt;text&gt;</a></strong></dt>

<dd>
<p>Enter the species name for the database to connect to. Common aliases, 
e.g. human, may be acceptable. Check the Ensembl website for the species 
databases available.</p>
</dd>
<dt><strong><a name="out_filename" class="item">--out &lt;filename&gt;</a></strong></dt>

<dd>
<p>Specify the output filename. By default, it uses the species name. An 
extension will be added if necessary.</p>
</dd>
<dt><strong><a name="no_chromo" class="item">--(no)chromo</a></strong></dt>

<dd>
<p>Boolean flag to indicate whether or not to write (or not) features for 
the toplevel chromosomes/scaffolds/contigs/sequences in the database. 
The GFF type is the generic term 'chromosome'. The default is true.</p>
</dd>
<dt><strong><a name="no" class="item">--no(protein)</a></strong></dt>

<dd>
<p>Boolean flag to indicate whether or not to collect protein-coding genes 
from the database. The default is true.</p>
</dd>
<dt><strong><a name="no_rna" class="item">--(no)rna</a></strong></dt>

<dd>
<p>Boolean flag to indicate whether or not to collect non-coding RNA genes, 
including misc_RNA, snRNA, snoRNA, rRNA, tRNA, miRNA, and piRNA, from 
the database. The default is false.</p>
</dd>
<dt><strong><a name="host_host_address" class="item">--host &lt;host.address&gt;</a></strong></dt>

<dd>
<p>Specify the Internet address of the Ensembl public MySQL database host. 
The default value is 'ensembldb.ensembl.org'.</p>
</dd>
<dt><strong><a name="user_text" class="item">--user &lt;text&gt;</a></strong></dt>

<dd>
<p>Specify the user name to connect as to the public database. 
The default value is 'anonymous'.</p>
</dd>
<dt><strong><a name="help" class="item">--help</a></strong></dt>

<dd>
<p>Display this POD documentation.</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program will connect to the public Ensembl MySQL database and 
retrieve the latest genome annotation for a given species. It will 
generate a GFF3 annotation file suitable for loading into a Bio::DB 
database or genome browser. The GFF3 features generated are 
multi-level nested gene-&gt;mRNA-&gt;CDS features. It will optionally 
generate features for the top-level sequences (chromosomes, contigs, 
scaffolds, etc.) and non-coding RNA genes (snRNA, tRNA, rRNA, etc.).</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
