<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>map_data.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>map_data.pl</p>
<p>A script to map data relative to and flanking a genomic feature</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<pre>

 map_data.pl --db &lt;database&gt; --feature &lt;name&gt; --out &lt;file&gt; [--options]
 map_data.pl --in &lt;in_filename&gt; --out &lt;out_filename&gt; [--options]</pre>
<pre>

  Options:
  --db &lt;database&gt;
  --feature [type, type:source]
  --in &lt;filename&gt; 
  --out &lt;filename&gt;
  --data &lt;dataset_name&gt;
  --value [score | length | count | enumerate]
  --method [mean | median | stddev | sum]
  --win &lt;integer&gt;
  --num &lt;integer&gt;
  --pos [5 | 3 | m]
  --strand [sense | antisense | all]
  --set_strand
  --(no)sum
  --(no)smooth
  --(no)log
  --help</pre>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="db_database" class="item">--db &lt;database&gt;</a></strong></dt>

<dd>
<p>Specify the name of the BioPerl GFF database to use. This is optional
if the desired database is defined in the input file metadata.</p>
</dd>
<dt><strong><a name="out_filename" class="item">--out &lt;filename&gt;</a></strong></dt>

<dd>
<p>Specify the output file name. Required for new files; otherwise, 
input files will be overwritten unless specified.</p>
</dd>
<dt><strong><a name="in_filename" class="item">--in &lt;filename&gt;</a></strong></dt>

<dd>
<p>Specify the filename of a data table containing the list of 
features to map nucleosomes around. The file must be in the
'tim_data' format and specify a feature to use. If an input 
file is not specified, then a new list of features will be 
generated from the database.</p>
</dd>
<dt><strong><a name="feature_type_type_source" class="item">--feature [type, type:source]</a></strong></dt>

<dd>
<p>Specify the type of feature to map data around. The feature may be 
listed either as GFF type or GFF type:source. The list 
of features will be automatically generated from the database.</p>
<p>This is optional if the features are defined in the input file.</p>
</dd>
<dt><strong><a name="data_name" class="item">--data &lt;name&gt;</a></strong></dt>

<dd>
<p>Specify the name of the microarray data set for which you wish to 
collect data. If not specified, the data set may be chosen
interactively from a presented list. Alternatively, other
features may be collected, and should be specified using the type 
(GFF method:source), especially when collecting alternative data values.</p>
</dd>
<dt><strong><a name="value_score_length_count_enumerate" class="item">--value [score | length | count | enumerate]</a></strong></dt>

<dd>
<p>Specify the type of data value or attribute to collect from the dataset  
or features for each region. By default, the score is collected. 
Alternatively, a count of the number of features present in the window, or 
the length of the features, may be collected.</p>
</dd>
<dt><strong><a name="method_mean_median_sum_stddev" class="item">--method [mean | median | sum | stddev]</a></strong></dt>

<dd>
<p>Specify the method of combining multiple values within each window. The mean, 
median, sum, or standard deviation of the values may be reported. The default 
value is mean for score and length values, or sum for enumerated values.</p>
</dd>
<dt><strong><a name="win_integer" class="item">--win &lt;integer&gt;</a></strong></dt>

<dd>
<p>Specify the window size. Default 50</p>
</dd>
<dt><strong><a name="num_integer" class="item">--num &lt;integer&gt;</a></strong></dt>

<dd>
<p>Specify the number of windows on either side of the feature position 
(total number will be 2 x [num]). Default 20</p>
</dd>
<dt><strong><a name="pos_5_3_m" class="item">--pos [ 5 | 3 | m ]</a></strong></dt>

<dd>
<p>Indicate the relative position of the feature around which the 
data is mapped. Three values are accepted: &quot;5&quot; indicates the 
5' prime end is used, &quot;3&quot; indicates the 3' end is used, and &quot;m&quot; or &quot;1&quot;
indicates the middle of the feature is used. The default is to 
use the 5' end, or the start position of unstranded features. 
If the feature &quot;tts&quot; is selected above, the 3' end is 
automatically selected.</p>
</dd>
<dt><strong><a name="strand_sense_antisense_all" class="item">--strand [sense | antisense | all]</a></strong></dt>

<dd>
<p>Specify whether stranded data should be collected for each of the 
datasets. Either sense or antisense (relative to the feature) data 
may be collected. The default value is 'all', indicating all 
data will be collected.</p>
</dd>
<dt><strong><a name="set_strand" class="item">--set_strand</a></strong></dt>

<dd>
<p>For features that are not inherently stranded (strand value of 0), 
impose an artificial strand for each feature (1 or -1). This will 
have the effect of enforcing a relative orientation for each feature, 
or to collected stranded data. This requires the presence a 
column in the input data file with a name of &quot;strand&quot;. Hence, it 
will not work with newly generated datasets, but only with input 
data files. Default is false.</p>
</dd>
<dt><strong><a name="no_sum" class="item">--(no)sum</a></strong></dt>

<dd>
<p>Indicate that the data should be averaged across all features at
each position, suitable for graphing. A separate text file will 
be written with the suffix '_summed' with the averaged data. 
Default is true.</p>
</dd>
<dt><strong><a name="no_smooth" class="item">--(no)smooth</a></strong></dt>

<dd>
<p>Indicate that windows without values should (not) be interpolated
from neighboring values.</p>
</dd>
<dt><strong><a name="no_log" class="item">--(no)log</a></strong></dt>

<dd>
<p>Dataset values are (not) in log2 space and should be treated 
accordingly. Output values will be in the same space.</p>
</dd>
<dt><strong><a name="help" class="item">--help</a></strong></dt>

<dd>
<p>Display this help</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program will map data around a genomic feature. Features
may include the transcription start site (TSS) of a transcript, tRNA gene,
or some other genomic feature. The data is collected in a series of windows
flanking the feature start, end, or midpoint position. The number and size of
windows are specified on the command line, or the program will default to
20 windows (on either side of the feature, 40 total) of 50 bp size. These
default window settings corresponds to 1 kb on either side of the feature.
Windows without a value may be interpolated (smoothed) from neigboring 
values, if available.</p>
<p>The default value that is collected is a dataset score (e.g. microarray 
values). However, other values may be collected, including 'count' or 
'length'. Use the --method argument to collect alternative values.</p>
<p>The program writes out a tim data formatted text file.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112

</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
