<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>map_transcripts.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>map_transcripts.pl</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>map_transcripts.pl --ffile &lt;filename&gt; --rfile &lt;filename&gt; --out &lt;filename&gt;</p>
<p>map_transcripts.pl --db &lt;name&gt; --win &lt;i&gt; --step &lt;i&gt; --fdata &lt;name&gt; 
--rdata &lt;name&gt; --thresh &lt;i&gt; --out &lt;filename&gt;</p>
<pre>

  Options:
  --out &lt;filename&gt;
  --ffile &lt;filename&gt;
  --rfile &lt;filename&gt;
  --db &lt;db_name&gt;
  --win &lt;integer&gt;
  --step &lt;integer&gt;
  --thresh &lt;number&gt;
  --fdata &lt;dataset&gt;
  --rdata &lt;dataset&gt;
  --tol &lt;integer&gt;
  --min &lt;integer&gt;
  --(no)gff
  --source &lt;text&gt;
  --help</pre>
<pre>
 
=head1 OPTIONS</pre>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="out_filename" class="item">--out &lt;filename&gt;</a></strong></dt>

<dd>
<p>Provide the base output filename for the transcript data file, GFF file, 
and summary file.</p>
</dd>
<dt><strong><a name="ffile_filename" class="item">--ffile &lt;filename&gt;</a></strong></dt>

<dt><strong><a name="rfile_filename" class="item">--rfile &lt;filename&gt;</a></strong></dt>

<dd>
<p>Provide the names of eniched regions files representing transcription 
fragments. The files should be generated by the program 
<code>find_enriched_regions.pl</code>. If these files are not provided, they 
will be automatically generated by running the find_enriched_regions.pl 
program using the parameters defined here. Two files should be provided: 
each representing transcription fragments on the forward and reverse 
strands.</p>
</dd>
<dt><strong><a name="db_database_name" class="item">--db &lt;database_name&gt;</a></strong></dt>

<dd>
<p>Specify a Bioperl Bio::DB::SeqFeature::Store database. Required for 
generating the transcription fragment enriched_regions files. It may 
be gleaned from the metadata of the provided enriched_region files.</p>
</dd>
<dt><strong><a name="win_integer" class="item">--win &lt;integer&gt;</a></strong></dt>

<dd>
<p>Specify the window size in bp to scan the genome for enriched regions. 
When the find_enriched_regions.pl program is run, trimming is turned 
on so that the final regions are not limited to multiples of 
window size.</p>
</dd>
<dt><strong><a name="step_integer" class="item">--step &lt;integer&gt;</a></strong></dt>

<dd>
<p>Specify the step size in bp for advancing the window across the genome 
when scanning for enriched regions.</p>
</dd>
<dt><strong><a name="thresh_number" class="item">--thresh &lt;number&gt;</a></strong></dt>

<dd>
<p>Specify the threshold value to identify regions that are enriched, 
i.e. expressed. A relatively low value should be set to ensure 
low expressed transcripts are also included. The value may be gleaned 
from the metadata in the provided enriched regions files.</p>
</dd>
<dt><strong><a name="fdata_dataset" class="item">--fdata &lt;dataset&gt;</a></strong></dt>

<dt><strong><a name="rdata_dataset" class="item">--rdata &lt;dataset&gt;</a></strong></dt>

<dd>
<p>Specify the name of the datasets in the database representing the 
individual forward and reverse transcription data. Usually this 
data is provided separately for each strand; a single dataset with 
both strands of data is not accepted. These may be gleaned from the 
metadata of the provided enriched regions files.</p>
</dd>
<dt><strong><a name="tol_integer" class="item">--tol &lt;integer&gt;</a></strong></dt>

<dd>
<p>Specify the tolerance distance in bp with which the ends of a 
transcription fragment and gene need to be within to call the 
transcription fragment a complete transcript. If the ends 
exceed the tolerance value, then the transcription fragment is 
labeled as 'overlap'. The default value is 20 bp.</p>
</dd>
<dt><strong><a name="min_integer" class="item">--min &lt;integer&gt;</a></strong></dt>

<dd>
<p>Small, spurious transcription fragments may be identified, particularly 
if the threshold is set too low or the dataset is noisy. This 
parameter sets the minimum size of the transcription fragment to 
initiate a search for associated known genes. The default value is 
100 bp.</p>
</dd>
<dt><strong><a name="no_gff" class="item">--(no)gff</a></strong></dt>

<dd>
<p>Indicate whether a GFF file should also be written. A version 3 file 
is generated. The default is true.</p>
</dd>
<dt><strong><a name="source_text" class="item">--source &lt;text&gt;</a></strong></dt>

<dd>
<p>Specify the GFF source value when writing a GFF file. The default 
value is the name of this program.</p>
</dd>
<dt><strong><a name="help" class="item">--help</a></strong></dt>

<dd>
<p>Display the POD documentation.</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program will identify transcription fragments that correspond to 
gene transcripts using available transcriptome microarray data.</p>
<p>Transcription fragments are identified as windows of enrichment using the 
script 'find_enriched_regions.pl'. This script can either be automatically
executed using the specified parameters, or run separately. If run separately, 
the output files should be indicated (--ffile and --rfile).</p>
<p>Each enriched window, or transcription fragment, is checked for corresponding 
or overlapping genomic features. The name(s) of the overlapping gene(s) are 
reported, as well as a classification for the transcript. Transcripts that 
completely contain (within a 20 bp tolerance) a single known gene (ORF or 
ncRNA) are labeled as 'complete'. Transcription fragments that simply overlap a
known gene are labeled as 'overlap'. Transcription fragments that overlap more 
than one annotated gene are labeled as 'multi-orf'.</p>
<p>Transcription fragments that only overlap annotated genes on the opposite 
strand are labeled as 'anti-sense'. Transcription fragments overlapping 
repetitive elements or no known feature are also reported.</p>
<p>The program does not attempt to identify individual exons and introns. 
Rather, it is primarily focused on only identifying transcription start 
and stop sites. All transcription fragments which overlap a known 
known gene are considered to be part of the same of transcript and are 
therefore merged together into one transcript.</p>
<p>Some rudimentary calculations are performed to identify the length of 
the 5' and 3' UTRs.</p>
<p>The program writes out a tab delimited text file. It will also write out
a gff file for the genome browser. It also writes out a summary report file.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
