<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>my_gff2gff3.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>my_gff2gff3.pl</p>
<p>A script to convert my data GFF v.2 files to GFF3</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>my_gff2gff3.pl [--options ...] &lt;file1.gff&gt; &lt;file2.gff&gt; ...</p>
<pre>

  --type &lt;text&gt;
  --source &lt;text&gt;
  --mt &lt;text&gt;
  --help</pre>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="type_text" class="item">--type &lt;text&gt;</a></strong></dt>

<dd>
<p>Provide a text string to bue used in the method/class/type
column. Since we are generating a v.3 GFF file, the type 
value should map to a valid Sequence Ontology Term. Common 
data terms include 'microarray_oligo', 'tag', or 'STS'. See
<a href="http://www.sequenceontology.org">http://www.sequenceontology.org</a> for more information. 
Default is to use the original.</p>
</dd>
<dt><strong><a name="source_text" class="item">--source &lt;text&gt;</a></strong></dt>

<dd>
<p>Optionally provide a new text string to replace the 'source' 
field. Default is to use the original.</p>
</dd>
<dt><strong><a name="mt_text" class="item">--mt &lt;text&gt;</a></strong></dt>

<dd>
<p>I used to rename the mitochrondrial chromosome in cerevisiae to 
'chr17' to help facilitate analysis. This was not very 
professional. A new name may be provided for renaming all 'chr17' 
features (back) to the new (original) name.</p>
</dd>
<dt><strong><a name="help" class="item">--help</a></strong></dt>

<dd>
<p>Display the POD documentation</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program will convert a simple data version 2 GFF formatted file into a GFF3 
file. Typical GFF data files that I had generated are in GFF v.2 format. 
Additionally, I used the 'type' or 'method' column (the third column) as 
a means of identifying and grouping values into a single data set. These 
were further often identified by the source tag of 'data'. Finally, the 
group field (ninth column) was simply in the format 'Experiment $type'.</p>
<p>With bioperl, GBrowse, and other programs moving towards format v.3 
(GFF3), these practices are no longer acceptable. This program attempts 
to convert my GFF v.2 data files into valid GFF3 files. First, the type 
column must be a valid Sequence Ontology term. For microarray data, this 
can be 'microarray_oligo'. For ChIP-seq data, the best terms may either be 
'tag' or 'STS' (for Sequence Tag Site). Next, the group field should 
contain both Name and ID tags. The old 'type' field will become the new 
Name tag and will be used for grouping. The ID tag must be made unique 
(accomplished by appending a unique number to the Name). Finally, the 
source field could be better utilized with more unique information.</p>
<p>The program will accept and process one or more gff input files. It will 
write a new file, renaming the extension to '.gff3' for each input file.</p>
<p>NOTE that this program will not deal with parent-child relationships. It 
is strictly designed to work with very simple one-line features, such as 
representing microarray data.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
