#summary Converted POD documentation.
#labels ManPage
_This page was generated from POD using [http://code.google.com/p/pod2gcw pod2gcw] and is not intended for editing._

==NAME==
get_ensembl_annotation.pl

==SYNOPSIS==
get_ensembl_annotation.pl `[`--options...`]` --species <**text>

{{{
  Options:
  --species <text>
  --out <filename>
  --(no)chromo
  --(no)protein
  --(no)rna
  --host <host.address>
  --user <text>
  --help
}}}
{{{
}}}
==OPTIONS==
The command line flags and descriptions:

__--species <**text>

__
  Enter the species name for the database to connect to. Common aliases,  e.g. human, may be acceptable. Check the Ensembl website for the species  databases available.
__
  __
__--out <**filename>

__
  Specify the output filename. By default, it uses the species name. An  extension will be added if necessary.
__
  __
__--(no)chromo

__
  Boolean flag to indicate whether or not to write (or not) features for  the toplevel chromosomes/scaffolds/contigs/sequences in the database.  The GFF type is the generic term 'chromosome'. The default is true.
__
  __
__--no(protein)

__
  Boolean flag to indicate whether or not to collect protein-coding genes  from the database. The default is true.
__
  __
__--(no)rna

__
  Boolean flag to indicate whether or not to collect non-coding RNA genes,  including misc_RNA, snRNA, snoRNA, rRNA, tRNA, miRNA, and piRNA, from  the database. The default is false.
__
  __
__--host <**host.address>

__
  Specify the Internet address of the Ensembl public MySQL database host.  The default value is 'ensembldb.ensembl.org'.
__
  __
__--user <**text>

__
  Specify the user name to connect as to the public database.  The default value is 'anonymous'.
__
  __
__--help

__
  Display this POD documentation.
__
  __
==DESCRIPTION==
This program will connect to the public Ensembl MySQL database and  retrieve the latest genome annotation for a given species. It will  generate a GFF3 annotation file suitable for loading into a Bio::DB  database or genome browser. The GFF3 features generated are  multi-level nested gene->mRNA->CDS features. It will optionally  generate features for the top-level sequences (chromosomes, contigs,  scaffolds, etc.) and non-coding RNA genes (snRNA, tRNA, rRNA, etc.).

{{{
}}}
{{{
}}}
==AUTHOR==
{{{
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112
}}}
This package is free software; you can redistribute it and/or modify it under the terms of the GPL (either version 1, or at your option, any later version) or the Artistic License 2.0.

{{{
}}}
