#summary Converted POD documentation.
#labels ManPage
_This page was generated from POD using [http://code.google.com/p/pod2gcw pod2gcw] and is not intended for editing._

==NAME==
find_enriched_regions.pl

==SYNOPSIS==
{{{
 find_enriched_regions.pl --db <db_name> [--options]
 }}}
{{{
  Options:
  --db <db_name>
  --data <dataset>
  --out <filename>
  --win <integer>
  --step <integer>
  --tol <integer>
  --thresh <number>
  --sd <number>
  --median
  --deplete
  --(no)trim
  --(no)sort
  --(no)feat
  --(no)log
  --genes
  --gff
  --gz
  --help
}}}
{{{
 }}}
==OPTIONS==
The command line flags and descriptions:

__--db <**database_name>

__
  Specify a Bioperl Bio::DB::!SeqFeature**::Store database. Required.
__
  __
__--data <**dataset>

__
  Specify the name of the dataset from which to collect the scores.  If not specified, then the data set may be chosen interactively  from a presented list.
__
  __
__--out <**filename>

__
  Specify the output file name. If not specified, then it will be  automatically generated from dataset, window, step, and threshold  values.
__
  __
__--win <**integer>

__
  Specify the genomic bin window size in bp. Default value is 250 bp.
__
  __
__--step <**integer>

__
  Specify the step size for moving the window. Default value is  equal to the window size.
__
  __
__--tol <**integer>

__
  Specify the tolerance distance when merging windows together.  Windows not actually overlapping but within this tolerance  distance will actually be merged. Default value is 1/2 the  window size.
__
  __
__--thresh <**number>

__
  Specify the window score threshold explicitly rather than calculating a threshold based on standard deviations from the mean.
__
  __
__--sd <**number>

__
  Specify the multiple of standard deviations above the mean as a window score threshold. Default is 1.5 standard deviations. Be  quite careful with this method as it attempts to pull all of the  datapoints out of the database to calculate the mean and  standard deviation - which may be acceptable for limited tiling  microarrays but not acceptable for next generation sequencing  data with single bp resolution.
__
  __
__--median

__
  Indicate that a median score value should be calculated within each window rather than the default mean when determining whether the  window exceeds the threshold.
__
  __
__--deplete

__
  Specify whether depleted regions should be reported instead.  For example, windows whose scores are 1.5 standard deviations  below the mean, rather than above.
__
  __
__--(no)trim

__
  Indicate that the merged windows should (not) be trimmed of below  threshold scores on the ends of the window. Normally when windows  are merged, there may be some data points on the ends of the  windows whose scores don't actually pass the threshold, but were  included because the entire window mean (or median) exceeded  the threshold. This step removes those data points. The default  behavior is true.
__
  __
__--(no)feat

__
  Indicate that features overlapping the windows should be  identified. The default behavior is false.
__
  __
__--genes

__
  Write out a text file containing a list of the found overlapping genes.
__
  __
__--gff

__
  Indicate that a GFF version 3 file should be written out in addition to the data text file.
__
  __
__--(no)log

__
  Flag to indicate that source data is (not) log2, and to calculate  accordingly and report as log2 data.
__
  __
__--gz

__
  Compress the output file through gzip.
__
  __
__--help

__
  Display the POD documentation.
__
  __
==DESCRIPTION==
This program will search for regions in the genome that are enriched for a  particular data set. It walks through each chromosome using a  window of specified size (default 500 bp) and specified step size (default 100 bp). Data scores within a window that exceed a determined threshold will be noted. Adjacent windows are merged and then trimmed on the ends to the minimum thresholded window.

The threshold scores for identifying an enriched region may either be  explicitly set or automatically determined from the mean and standard  deviation (SD) of the entire collection of datapoints across the genome.  This, of course, assumes a normal distribution of datapoint scores, which may  or may not be suitable for the particular dataset. Note that the  automatic method may not appropriate for very extremely large datasets  (e.g. next generation sequencing) as it attempts to calculate the mean and SD  on all of the datapoints in the database.

The program writes out a tim data formatted text file consisting of chromosome,  start, stop, score, and overlapping gene or non-gene genomic features. It  will optionally write a GFF file.

{{{
}}}
==AUTHOR==
{{{
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112
}}}
This package is free software; you can redistribute it and/or modify it under the terms of the GPL (either version 1, or at your option, any later version) or the Artistic License 2.0.

{{{
}}}
{{{
}}}
{{{
}}}
{{{
}}}
