<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>split_bam_by_isize.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>split_bam_paired_by_size.pl</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>split_bam_paired_by_size.pl [--options] &lt;file.bam&gt;</p>
<pre>

  Options:
  --in &lt;file.bam&gt;
  --min &lt;integer&gt;
  --max &lt;integer&gt;
  --size &lt;min-max&gt;
  --out &lt;filename&gt;
  --at
  --help</pre>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="in_file_bam" class="item">--in &lt;file.bam&gt;</a></strong></dt>

<dd>
<p>Specify the file name of a binary BAM file as described for Samtools. 
Use samtools to convert text sam files to binary bam files. The file 
should be indexed; this program should be able to do it for you 
automatically if it is not (assuming the bam directory is writeable).</p>
</dd>
<dt><strong><a name="min_integer" class="item">--min &lt;integer&gt;</a></strong></dt>

<dd>
<p>Optionally specify the minimum size of fragment to include in the output
GFF file. The default value is 100 bp.</p>
</dd>
<dt><strong><a name="max_integer" class="item">--max &lt;integer&gt;</a></strong></dt>

<dd>
<p>Optionally specify the maximum size of fragment to include in the output
GFF file. The default value is 200 bp.</p>
</dd>
<dt><strong><a name="size_min_max" class="item">--size &lt;min-max&gt;</a></strong></dt>

<dd>
<p>When multiple size ranges are desired, they may be specified using the 
size option. Define the minimum and maximum size as a range separated by 
a dash (no spaces). Use this option repeatedly for multiple size ranges. 
The size option takes precedence over the min and max options.</p>
</dd>
<dt><strong><a name="out_filename" class="item">--out &lt;filename&gt;</a></strong></dt>

<dd>
<p>Optionally specify the base name of the output file. The default is to use 
base input name. The output file names are appended with '.$min_$max'.</p>
</dd>
<dt><strong><a name="at" class="item">--at</a></strong></dt>

<dd>
<p>Boolean option to indicate that only fragments whose paired sequence reads 
end in a [AT] nucleotide should be included in the output GFF file. 
Micrococcal nuclease (MNase) cuts (almost?) exclusively at AT dinucleotides; 
this option ensures that the fragment is more likely derived from a MNase 
cut.</p>
</dd>
<dt><strong><a name="help" class="item">--help</a></strong></dt>

<dd>
<p>Display the POD documentation</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program will read a BAM file containing paired-read alignments and
write a new BAM file containing only those successfully aligned paired 
reads whose insert size fall within the set minimum and maximum 
lengths inclusively (defaults of 100 and 200 bp, respectively).</p>
<p>Multiple size ranges may be specified, and pairs within each range are 
written to separate files.</p>
<p>Additionally, fragments may be checked for the presence of an A/T nucleotide 
at the 5' end of the sequence, as is usually produced by digestion with 
Micrococcal Nuclease.</p>
<p>The input and output files are BAM files as described by the Samtools 
project (http://samtools.sourceforge.net).</p>
<p>The input file should be sorted and indexed prior to sizing. The output file 
will also be automatically re-sorted and re-indexed for you.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
