<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>bin_genomic_data.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>bin_genomic_data.pl</p>
<p>A script to bin genomic data into windows</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<pre>

 bin_genomic_data.pl --data &lt;filename&gt; --in &lt;filename&gt; --method [method]</pre>
<pre>

 bin_genomic_data.pl --data &lt;filename&gt; --new --win &lt;integer&gt; --db &lt;db_name&gt;
     --out &lt;filename&gt; --method [method]</pre>
<pre>

  Options:
  --data &lt;filename&gt;
  --in &lt;filename&gt;
  --new
  --method [enumerate | mean | median]
  --out &lt;filename&gt;
  --db &lt;database_name&gt;
  --win &lt;integer&gt;
  --step &lt;integer&gt;
  --paired
  --span
  --midpoint
  --shift &lt;integer&gt;
  --interbase
  --interpolate
  --(no)log
  --gff
  --(no)gz
  --help</pre>
<pre>
 
=head1 OPTIONS</pre>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="data_filename" class="item">--data &lt;filename&gt;</a></strong></dt>

<dd>
<p>Specify the source genomic data file containing the datapoints to be 
binned. Supported files include .gff, .bed, , .wig, .sgr, and .bam files. 
Text files may be gzipped.</p>
</dd>
<dt><strong><a name="in_filename" class="item">--in &lt;filename&gt;</a></strong></dt>

<dd>
<p>Specify the filename of an existing data table of genomic bins. It should 
be in the tim data file format with metadata and have the feature 'genome'.
It may be generated with this program or 'get_datasets.pl'. Required 
unless --new is used.</p>
</dd>
<dt><strong><a name="new" class="item">--new</a></strong></dt>

<dd>
<p>Indicate that a new data file should be generated. Required unless --in 
is used.</p>
</dd>
<dt><strong><a name="method_enumerate_mean_median" class="item">--method [enumerate | mean | median]</a></strong></dt>

<dd>
<p>Indicate which method should be used to collect the binned data. 
Acceptable values include:</p>
<pre>

 - enumerate     Sum the number of unique features in the bin.
 - mean          Take the mean score of features in the bin.
 - median        Take the median score of features in the bin.</pre>
</dd>
<dt><strong><a name="out_filename" class="item">--out &lt;filename&gt;</a></strong></dt>

<dd>
<p>Specify the output file name. Required if generating a new binned 
data file, otherwise if an input file was specified it will be 
overwritten.</p>
</dd>
<dt><strong><a name="db_database_name" class="item">--db &lt;database_name&gt;</a></strong></dt>

<dd>
<p>Specify a database. This is used only to retrieve chromosome 
information to generate genomic bins. Required if --new is used.</p>
</dd>
<dt><strong><a name="win_integer" class="item">--win &lt;integer&gt;</a></strong></dt>

<dd>
<p>Specify the genomic bin window size in bp. Required if --new is used.</p>
</dd>
<dt><strong><a name="step_integer" class="item">--step &lt;integer&gt;</a></strong></dt>

<dd>
<p>Specify the step size for moving the window when generating bins. 
The default is to equal the window size.</p>
</dd>
<dt><strong><a name="paired" class="item">--paired</a></strong></dt>

<dd>
<p>Indicate that the data bam file consists of paired-end alignments 
rather than single-end alignments.</p>
</dd>
<dt><strong><a name="span" class="item">--span</a></strong></dt>

<dd>
<p>Assign feature value (or count) at each genomic bin across the 
feature. This is dependent on the metadata step value when the 
genomic bins were generated.</p>
</dd>
<dt><strong><a name="midpoint" class="item">--midpoint</a></strong></dt>

<dd>
<p>Assign the the feature value (or count) at the feature's midpoint 
rather than the default feature's start point.</p>
</dd>
<dt><strong><a name="shift_integer" class="item">--shift &lt;integer&gt;</a></strong></dt>

<dd>
<p>For stranded single-end alignments from a bam data file, the start 
position may be shifted in the 3' direction by the indicated 
number of bp. This is to account for ChIP-Seq data where the peak 
of tag counts is offset from the actual center of the sequenced 
fragments. Use a shift value of 1/2 the mean fragment length of the 
sequencing library.</p>
</dd>
<dt><strong><a name="interbase" class="item">--interbase</a></strong></dt>

<dd>
<p>Source data is in interbase coordinate (0-base) system. Shift the 
start position to base coordinate (1-base) system. This only affects 
BED and SGR source files, which may or may not be in interbase format 
(the others normally are not if they follow specifications). Default 
is false.</p>
</dd>
<dt><strong><a name="interpolate" class="item">--interpolate</a></strong></dt>

<dd>
<p>Flag to indicate that those bins lacking data should be interpolated 
from neighboring bins. Only 'mean' or 'median' collected values are 
interpolated. (we assume you don't want to interpolate enumerated 
counts, right????).</p>
</dd>
<dt><strong><a name="no_log" class="item">--(no)log</a></strong></dt>

<dd>
<p>Flag to indicate that source data is (not) log2, and to calculate 
accordingly and report as log2 data.</p>
</dd>
<dt><strong><a name="gff" class="item">--gff</a></strong></dt>

<dd>
<p>Write a new gff output file instead of a normal tim data file.</p>
</dd>
<dt><strong><a name="gz" class="item">--gz</a></strong></dt>

<dd>
<p>Compress the output file through gzip.</p>
</dd>
<dt><strong><a name="help" class="item">--help</a></strong></dt>

<dd>
<p>Display the POD documentation.</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program will convert genomic data into genomic binned data. The genomic 
bins are generated from the indicated database as segments of the genome of 
the specified window size. The step size may be optionally specified, or 
defaults to the window size. The source data may be collected from multiple 
sources, including GFF, BED, BAM, WIG, or SGR files. The data may be 
collected in one of two ways.</p>
<p>First, the program can enumerate (or count) features which overlap with a 
genomic bin. Sequence tag (BAM file) alignments (both single- and 
paired-end), BED, and GFF features may be counted. The feature start point, 
or optionally the midpoint, is used to record the feature. Alternatively, the 
counts may be enumerated across the entire span of the feature when using 
small (single bp) bins.</p>
<p>Second, the program can statistically combine the scores (values) of features
that overlap each genomic bin. Feature scores may be collected from GFF, BED, 
WIG, or SGR files. Either the mean or the median value can be recorded for the 
genomic bin.</p>
<p>Genomic bins lacking a score may be interpolated from neighboring bins. Up to
four consecutive non-value bins may interpolated from neighboring bins that
contain values.</p>
<p>The program reads/writes a data file compatable with the 'get_datasets.pl' 
program. It can optionally write a new GFF file based on the genomic binned
data.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112

</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
