<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>data2wig.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>data2wig.pl</p>
<p>A script to convert a generic data file into a wig file.</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>data2wig.pl [--options...] &lt;filename&gt;</p>
<pre>

  Options:
  --in &lt;filename&gt;
  --out &lt;filename&gt; 
  --step [fixed,variable]
  --size &lt;integer&gt;
  --score &lt;column_index&gt;
  --name &lt;text&gt;
  --(no)track
  --inter
  --format [0,1,2,3]
  --(no)midpoint
  --bigwig|--bw
  --chromof &lt;filename&gt;
  --db &lt;database&gt;
  --(no)gz
  --help</pre>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="in_filename" class="item">--in &lt;filename&gt;</a></strong></dt>

<dd>
<p>Specify the file name of a data file, with or without the --in flag. 
The file may be any tab-delimited text file (preferably in the tim data 
format as described in <code>tim_file_helper.pm</code>), GFF, SGR, or BED file. 
Recognizeable genome coordinate columns should be present, including 
chromosome, start, and stop. Data files collected using the 'genome' 
windows feature are ideal. The file may be compressed with gzip.</p>
</dd>
<dt><strong><a name="out_filename" class="item">--out &lt;filename&gt;</a></strong></dt>

<dd>
<p>Optionally specify the name of of the output file. The track name is 
used as default. The '.wig' extension is automatically added if required.</p>
</dd>
<dt><strong><a name="step_fixed_variable" class="item">--step [fixed,variable]</a></strong></dt>

<dd>
<p>The type of step progression for the wig file. Two wig formats are available:
'fixedStep' where data points are positioned at equal distances along the 
chromosome, and 'variableStep' where data points are not equally spaced 
along the chromosome. The 'fixedStep' wig file has one column of data 
values (score), while the 'variableStep' wig file has two columns
(position and score). If the option is not defined, then the format is 
automatically determined from the metadata of the file.</p>
</dd>
<dt><strong><a name="size_integer" class="item">--size &lt;integer&gt;</a></strong></dt>

<dd>
<p>Optionally define the step size in bp for 'fixedStep' wig file. This 
value is automatically determined from the table's metadata, if available. 
If the --step option is explicitly defined as 'fixed', then the step size 
may also be explicitly defined. If this value is not explicitly
defined or automatically determined, the variableStep format is used by
default.</p>
</dd>
<dt><strong><a name="score_column_index" class="item">--score &lt;column_index&gt;</a></strong></dt>

<dd>
<p>Indicate the column index (0-based) of the dataset in the data table 
to be used for the score. If a GFF file is used as input, the score column is 
automatically selected. If not defined as an option, then the program will
interactively ask the user for the column index from a list of available
columns.</p>
</dd>
<dt><strong><a name="name_text" class="item">--name &lt;text&gt;</a></strong></dt>

<dd>
<p>The name of the track defined in the wig file. The default is to use 
the name of the chosen score column, or, if the input file is a GFF file, 
the base name of the input file.</p>
</dd>
<dt><strong><a name="no_track" class="item">--(no)track</a></strong></dt>

<dd>
<p>Do (not) include the track line at the beginning of the wig file. Wig 
files normally require a track line, but if you will be converting to 
the binary bigwig format, the converter requires no track line. Why it 
can't simply ignore the line is beyond me. This option is automatically 
set to false when the --bigwig option is enabled.</p>
</dd>
<dt><strong><a name="inter" class="item">--inter</a></strong></dt>

<dd>
<p>Source data is in interbase coordinate (0-base) system. Shift the 
start position to base coordinate (1-base) system. Wig files are by 
definition 1-based. Default is false.</p>
</dd>
<dt><strong><a name="no_midpoint" class="item">--(no)midpoint</a></strong></dt>

<dd>
<p>A boolean value to indicate whether the 
midpoint between the actual 'start' and 'stop' values
should be used. The default is to use only the 'start' position.</p>
</dd>
<dt><strong><a name="format_0_1_2_3" class="item">--format [0,1,2,3]</a></strong></dt>

<dd>
<p>Indicate the number of decimal places the score value should
be formatted. Acceptable values include 0, 1, 2, or 3 places.
The default is to not format the score value.</p>
</dd>
<dt><strong><a name="bigwig_bw" class="item">--bigwig --bw</a></strong></dt>

<dd>
<p>Indicate that a binary BigWig file should be generated instead of 
a text wiggle file. A .wig file is first generated, then converted to 
a .bw file, and then the .wig file is removed.</p>
</dd>
<dt><strong><a name="chromof_filename" class="item">--chromof &lt;filename&gt;</a></strong></dt>

<dd>
<p>When converting to a BigWig file, provide a two-column tab-delimited 
text file containing the chromosome names and their lengths in bp. 
Alternatively, provide a name of a database, below.</p>
</dd>
<dt><strong><a name="db_database" class="item">--db &lt;database&gt;</a></strong></dt>

<dd>
<p>Specify the database from which chromosome lengths can be derived when 
generating a bigwig file. This option is only required when generating 
bigwig files. It may also be supplied from the metadata in the source 
data file.</p>
</dd>
<dt><strong><a name="no_gz" class="item">--(no)gz</a></strong></dt>

<dd>
<p>A boolean value to indicate whether the output wiggle 
file should be compressed with gzip.</p>
</dd>
<dt><strong><a name="help" class="item">--help</a></strong></dt>

<dd>
<p>Display the POD documentation</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program will convert a data file into a wig formatted text
file. This presumes that the data file contains the chromosomal coordinates, 
i.e. chromosome, start, and (optionally) stop. If they are not found, the 
conversion will fail. The relevant columns should 
therefore be labeled 'chromosome', 'start', and 'stop' or 'end'.</p>
<p>The wig file format is specified by documentation supporting the UCSC 
Genome Browser and detailed at this location:
<a href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">http://genome.ucsc.edu/goldenPath/help/wiggle.html</a>
Two formats are supported, 'fixedStep' and 'variableStep'.</p>
<p>Alternatively, a binary BigWig file may be generated as opposed to a 
text wiggle file. The binary format is preferential to the text version 
for a variety of reasons, including fast, random access and no loss in 
data value precision. More information can be found at this location:
<a href="http://genome.ucsc.edu/goldenPath/help/bigWig.html.">http://genome.ucsc.edu/goldenPath/help/bigWig.html.</a> Conversion requires 
BigWig file support, supplied by the biotoolbox module 
<code>tim_db_helper::bigwig</code>.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
